### ISSUE TITLE
Feature request: improvements

### ISSUE THREAD

**Issue by piegamesde** [JOML-CI/JOML#137]
I use this library a lot (I mean _excessively_) and those are the things while I was working that annoyed me. (Okay, some are less an annoyance and more like "breaks my program and I can't fix it", but doesn't matter). So here's a list of things I'd like to change (sorted by relevance): 

1. Make `MemUtils` take the read-only version instead of the implementation where it doesn't mutate the object.
2. The interfaces (especially the Vector classes) mostly don't need write access. So the implementations from `VectorNX` should be default methods in `VectorNXc` instead. (I'm also fine with delegating it to an external class like done in `MemUtils`.)
3. Point 1 and 2 are to make it easier to create own implementations of the Vector classes (that's why it is an interface in the first place, isn't it?).
4. Implement a solid and working truly immutable version of the vector classes
5. Create `Rectanglei` and `AABBdi`. I miss them.
6. Make all flavours of a class (especially Vector and Rectangle/AABBd) equally featured. Why do some have a `div()` operator, others not?

Please tell me for each of those points if I or somebody else can implement them. I generally don't do commits just for them to be rejected, so I prefer to ask in advance.

**Comment by httpdigest** [comment]
I totally see your points and understand your arguments. JOML does not have immutable classes and now we're looking for ways to have them.

Here is what I wanted to be doing for a long time and what seems to become more and more a necessity as maintaining and extending the already huge JOML code base is going to be a bigger pain:

The "grand scheme" of my currently favored solution to have immutable classes and to solve a whole bunch of other issues (amongst them your point 5 and 6) is to use a templating/generator approach in order to:
- generate the different type-flavors of all vector classes
- generate mutable and immutable kinds of all vector classes
- generate the "modify this/copy" vs. "modify dest" method overloads automatically
- allow to benefit from improvements to the Java language/runtime (such as the vector API in project Valhalla) without manually editing code manually
- do not use Java interfaces anymore but generate explicit classes only, because it is hard to design an interface/view on a vector that has both a well-defined semantic and is actually amenable for subtyping without sacrificing performance or extensibility issues in any derived classes.
- avoid "forgetting" certain methods in the various flavors of a JOML type (your point 6)
- even generate sources for another language such as JavaScript/TypeScript or use a transpiler such as [JSmooth](http://jsmooth.sourceforge.net/) which looks decent

The question is just what to use as a templating language, which allows us to still write the actual code of a vector/matrix/etc. method by hand and translate it to different type-flavors and mutable/immutable versions and possibly other future aspects as well. The methods in JOML classes are not really complex syntacically (they use few Java language features). At least simple things like assignments and binary arithmetic instructions must be supported, and it has to have knowledge about the types of variables and expressions for necessary type coercions/casts to support all type-flavors of all JOML types.

This is where is see [JavaParser](https://github.com/javaparser/javaparser) and [JavaSymbolSolver](https://github.com/javaparser/javasymbolsolver) coming into play. Those two projects look pretty decent and I've developed a small AST transformation to generate the vector methods from a Java template, such as:
```Java
public Vector3 add(Vector3 v) {
	Vector3 res = result();
	vector_add(this, v, res);
	return res;
}
```
which would generate:
```Java
public Vector3fm add(Vector3fm v) {
	Vector3fm res = this;
	float x, y, z;
	x = this.x + v.x;
	y = this.y + v.y;
	z = this.z + v.z;
	res.x = x;
	res.y = y;
	res.z = z;
	return res;
}
public Vector3fm add(Vector3fm v, Vector3fm dest) {
	Vector3fm res = dest;
	float x, y, z;
	x = this.x + v.x;
	y = this.y + v.y;
	z = this.z + v.z;
	res.x = x;
	res.y = y;
	res.z = z;
	return res;
}
```
for a mutable float vector class and:
```Java
public Vector3im add(Vector3im v) {
	float x, y, z;
	x = this.x + v.x;
	y = this.y + v.y;
	z = this.z + v.z;
	Vector3im res = new Vector3im(x, y, z);
	return res;
}
```
for the immutable version.

The "primitives" used by the template language has to include some way of saying "this is the result vector" (which may be 'this' or 'dest' or a new immtuable vector instance) and vector instructions in order to "compile" them down to scalar arithmetic now and to a Java vector API later in Java 10 or 11 or whenever it arrives.
Also, there would be primitives for generating JavaDocs comments.

So, I'd like to start with Java as the template language, because it has simple syntactic elements which a good parser exists today, and use a meta-programming solution with JavaParser and JavaSymbolSolver to give any possible code generator an understanding of a method's code to be able to derive new code from it.

With this in mind, it'll be somewhat a waste of time to develop immutable classes by hand now.
However, I do see your necessity to have them in a timely manner. You can of course develop immutable classes for yourself, which you already did to an extent in a fork of JOML, and use them in your projects for now. But I'd rather not like to add more classes/code to JOML for the immutable classes, because I would rather want to focus on a generator approach.

**Comment by piegamesde** [comment]
How long will this take until the first published version? I'd like to have at least point 1) and 2) implemented so that if I really need it, I can implement an immutable version in a fork (I'd actually go for an extension library instead but that doesn't matter).

Your idea of generating the code sounds cool, but three things:

 - I don't know what design you are after, but please make it so that vector immutability is a compile-time and not a run-time attribute
 - n-dimensional vector arithmetic is a bonus :)
 - I suggest you to create a new issue where you can make all design decisions and progress public so that we can discuss it

I just started messing around with my own design. It doesn't need a code generator, but it makes heavy use of abstraction and generics to make this work and it probably won't fit the lightweight-and-performant style this library is popular for.

**Comment by httpdigest** [comment]
> How long will this take until the first published version?

Certainly not until after some half a year or so. I'm having less and less time to work on JOML and something like this needs to be done thoroughly. Sorry that this is somewhat being kicked off with your immutable vector enhancement proposals, but modifying/maintaining the JOML sources currently is already a huge pain because of the many aspects not being properly abstracted over, such as the vector/matrix element type and the backing data type used for matrix elements. The latter is the reason why there is even a separate Android version of JOML having diverged quite heavily from the master/jre version of it, and which I haven't touched in a while.
My hope is that using a code generation approach this (different backing data types of matrix/vector elements) is another factor in the big cross product that we can get rid of.

> I'd like to have at least point 1) and 2) implemented so that if I really need it, I can implement an immutable version in a fork

On 1): The problem with abstracting over the concrete vector implementation using the interface in MemUtil is that this would defeat the whole purpose of MemUtil (which by the way is an internal class and should not be used anyway). The MemUtil methods are used for accelerating certain memory copy operations by making explicit assumptions about the field/memory layout of the vector/matrix classes. So changing the method signatures to accept vector interface parameters would then require runtime type checks inside the methods, which also would need to be modified when another vector implementation (yours) arrives. The best approach there is to just copy the MemUtil methods taking your specific vector class as parameter type.

On 2): JOML currently does not make use of any Java source language features greater than 1.1. And I tried to stay away from any advanced features such as interface default methods, generics or lambdas, because the whole design can very quickly get out of hand/control, and I never felt like using any advanced Java language features would actually make anything better. Also, we never know who else wants to use JOML for whatever purpose and in whichever Java environment. Some may even use it in a Java Microedition environment which is Java 1.3, implementing BD-J (Blu-ray Disc Java) applications. So, I'd rather not up the runtime requirements in any way. I know there are solutions like Retrolambda to backport to earlier Java language versions again, but that's something not everyone wants to do. Most people like "clone from GitHub and off you go."

> I gave up on the own design. The lack of proper OO abstraction of numbers in Java makes this impossible -- I'd had to implement a custom Number system including math to deal with this

This really shows the drawbacks of Java's approach to parametric polymorphism, being type erasure, as opposed to the template specialization approach C# and C++ take. See what is even possible with glm in C++. So, take type erasure together with the lack of operator overloading, and we basically cannot abstract over any numeric types when they are used in any arithmetic expressions. And, of course, due to type erasure erasing to Object, we also have performance degradation due to wrapping and unwrapping.

So, in a sense, my preferred approach is to bypass Java's lack of template specialization and implement that ourselves via code generation. Just that the language to use for that does not have to have near the expressiveness of a generic parametric polymorphism, since we just want different component types for our vectors.

**Comment by piegamesde** [comment]
Half a year is a long time, so I'll implement what I need, but to a private repository unless someone else really needs that code too.

Is there any way to see how that project is going or to even help out? I still have a lot of free time (constantly decreasing though).

**Comment by agent6262** [comment]
I ran into the same issue needing a proper read-only / immutable versions of the classes. I will work on a pr to try and fix issues 1, 2, 4, 6.

**Comment by piegamesde** [comment]
@httpdigest Has any progress been made? Is there a WIP to have a look at?
